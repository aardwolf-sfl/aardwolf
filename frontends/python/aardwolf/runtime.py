import os
import sys
import copy

from .writer import Writer
from .constants import *

WRITER = None
TEST_RESULTS = None


def _init_if_needed():
    global WRITER
    if WRITER is None:
        WRITER = Writer(os.path.join(os.environ.get(
            DATA_DEST, os.getcwd()), 'aard.trace'))
        WRITER.write_str('AARD/D1')


def _init_test_results_if_needed():
    global TEST_RESULTS
    if TEST_RESULTS is None:
        TEST_RESULTS = open(os.path.join(os.environ.get(
            DATA_DEST, os.getcwd()), 'aard.result'), 'w')


def write_stmt(id):
    _init_if_needed()
    WRITER.write_token(TOKEN_STATEMENT)
    WRITER.write_u64(id[0])
    WRITER.write_u64(id[1])


def write_expr(result, id):
    _init_if_needed()
    write_stmt(id)
    return result


# Accessor tree is in "S-expression" form. Examples:
# a = foo -> write_value(value, [])
# a, b = foo -> write_value(value, [[], []])
# a, (b, c) = foo -> write_value(value, [[], [[], []]])
def write_value(value, accessor_tree=None):
    _init_if_needed()

    if accessor_tree is None:
        accessor_tree = []

    for v in unpack_values(value, accessor_tree):
        if isinstance(v, bool):
            # This check needs to be *before* int check because bool is a subtype of
            # int.
            WRITER.write_token(TOKEN_DATA_BOOL)
            WRITER.write_u8(int(v))

        elif isinstance(v, int):
            # We consider all int values to be of type i64 because we have no
            # information of their actual type. Using `bit_length` may cause
            # unnecessary type changes in the runtime. Integers have unlimited
            # precision, if the v does not fit into 8 bytes, it is unsupported
            # by Aardwolf.
            try:
                _ = v.to_bytes(8, byteorder=sys.byteorder)
                WRITER.write_token(TOKEN_DATA_I64)
                WRITER.write_i64(v)
            except OverflowError:
                WRITER.write_token(TOKEN_DATA_UNSUPPORTED)

        elif isinstance(v, float):
            if sys.float_info.max == 1.7976931348623157e+308:
                WRITER.write_token(TOKEN_DATA_F64)
                WRITER.write_f64(v)
            elif sys.float_info.max == 3.4028235e+38:
                WRITER.write_token(TOKEN_DATA_F32)
                WRITER.write_f32(v)
            else:
                # Something ridiculous is happening
                WRITER.write_token(TOKEN_DATA_UNSUPPORTED)

        # Not yet supported by the core:
        # elif v is None:
        #     WRITER.write_token(TOKEN_DATA_NULL)

        # else:
        #     WRITER.write_token(TOKEN_DATA_NAMED)
        #     WRITER.write_cstr(type(v).__name__)

        else:
            WRITER.write_token(TOKEN_DATA_UNSUPPORTED)

    return value


class NonSubscriptable:
    def __getitem__(self, key):
        return None


def unpack_values(value, tree):
    if len(tree) == 0:
        if isinstance(tree, tuple):
            # *rest unpacking. For now, assume unsupported data type for now.
            return [None]
        else:
            return [value]
    else:
        # Ensure that we can use value[index].
        if not hasattr(value, '__getitem__'):
            if hasattr(value, '__iter__'):
                # Iterators are problematic. We cannot iterate through them
                # since that would make the original program invalid. We could
                # theoretically make a deep copy, but that has some issues on
                # its own (recursive data structure, IO resources, etc.). One
                # possibility is to wrap the iterator into our class but we need
                # to make sure that it does not break things like `isinstance`,
                # `getattr`, etc.
                value = NonSubscriptable()
            else:
                value = NonSubscriptable()

        output = []
        for index, node in enumerate(tree):
            output.extend(unpack_values(value[index], node))

        return output


def aardwolf_iter(iter, id, accessors=None):
    return AardwolfIter(iter, id, accessors)


class AardwolfIter:
    def __init__(self, inner, id, accessors):
        self.inner_ = iter(inner)
        self.id_ = id
        self.accessors_ = accessors

    def __iter__(self):
        return self

    def __next__(self):
        value = next(self.inner_)
        write_stmt(self.id_)
        write_value(value, self.accessors_)
        return value


def write_external(external):
    _init_if_needed()
    WRITER.write_token(TOKEN_EXTERNAL)
    WRITER.write_cstr(external)


def write_test_status(name, passed):
    _init_test_results_if_needed()
    status = 'PASS' if passed else 'FAIL'
    TEST_RESULTS.write(f'{status}: {name}\n')
