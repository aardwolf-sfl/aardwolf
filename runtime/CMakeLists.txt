cmake_minimum_required(VERSION 3.13)
project(aardwolf_runtime)

# C99
set(CMAKE_C_STANDARD 99)
set(CMAKE_C_STANDARD_REQUIRED on)

message(STATUS "Build type: ${CMAKE_BUILD_TYPE} with compiler: ${CMAKE_C_COMPILER}")

set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall -Wextra -pedantic")

if (CMAKE_BUILD_TYPE STREQUAL "Debug")
    add_definitions(-DDEBUG_ENABLED)
    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Werror")
elseif (CMAKE_BUILD_TYPE STREQUAL "Release")
endif ()

add_library(aardwolf_runtime STATIC runtime.c)
