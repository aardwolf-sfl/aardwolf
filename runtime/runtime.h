#ifndef AARDWOLF_RUNTIME_H
#define AARDWOLF_RUNTIME_H

#include <stdint.h>

#define TOKEN_STATEMENT 0xff
#define TOKEN_EXTERNAL 0xfe
#define TOKEN_DATA_UNSUPPORTED 0x10
#define TOKEN_DATA_I8 0x11
#define TOKEN_DATA_I16 0x12
#define TOKEN_DATA_I32 0x13
#define TOKEN_DATA_I64 0x14
#define TOKEN_DATA_U8 0x15
#define TOKEN_DATA_U16 0x16
#define TOKEN_DATA_U32 0x17
#define TOKEN_DATA_U64 0x18
#define TOKEN_DATA_F32 0x19
#define TOKEN_DATA_F64 0x20
#define TOKEN_DATA_BOOL 0x21
#define TOKEN_DATA_NAMED 0x28
#define TOKEN_DATA_NULL 0x29

typedef uint64_t file_ref_t;
typedef uint64_t statement_ref_t;

// Log executed statement.
void aardwolf_write_statement(file_ref_t file_id, statement_ref_t stmt_id);

// Log external identifier. This is intended for differentiating individual test
// cases such that aardwolf can assign blocks of traces to these test cases
// and use this information along with test case status given later for the
// analysis.
void aardwolf_write_external(const char *external);

// Separated function for generating the header. It is called automatically in
// normal version of runtime. This should be only called when bare runtime is
// used and the file header must be generated explicitly.
void aardwolf_write_header();

// Only primitive types. It is the responsibility of the frontend to correctly
// write serialize complex types (e.g., arrays or structures).
//
// Before every data dump, there must be indication of what types they are.
// It cannot be done beforehand just once, because in dynamically-typed
// languages, the type of a variable can change.
void aardwolf_write_data_unsupported();
void aardwolf_write_data_i8(int8_t value);
void aardwolf_write_data_i16(int16_t value);
void aardwolf_write_data_i32(int32_t value);
void aardwolf_write_data_i64(int64_t value);
void aardwolf_write_data_u8(uint8_t value);
void aardwolf_write_data_u16(uint16_t value);
void aardwolf_write_data_u32(uint32_t value);
void aardwolf_write_data_u64(uint64_t value);
void aardwolf_write_data_f32(float value);
void aardwolf_write_data_f64(double value);
void aardwolf_write_data_bool(uint8_t value);
void aardwolf_write_data_named(const char *value);
void aardwolf_write_data_null();
// TODO: Others

#endif // AARDWOLF_RUNTIME_H
