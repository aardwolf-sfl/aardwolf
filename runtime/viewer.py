import sys
import os
import struct

input_file = sys.argv[-1]

HEADER = b'AARD1\n'
TOKEN_STATEMENT = b'\xff'
TOKEN_EXTERNAL = b'\xfe'
TOKEN_DATA_I32 = b'\x11'
TOKEN_DATA_I64 = b'\x12'
TOKEN_DATA_F32 = b'\x15'
TOKEN_DATA_F64 = b'\x16'

def read_stmt(f):
    id = struct.unpack('Q', f.read(8))[0]
    return f'#{id}'

def read_i32(f):
    return struct.unpack('i', f.read(4))[0]

def read_i64(f):
    return struct.unpack('q', f.read(8))[0]

def read_f32(f):
    return struct.unpack('f', f.read(4))[0]

def read_f64(f):
    return struct.unpack('d', f.read(8))[0]

def read_str(f):
    result = ''
    byte = f.read(1)
    while byte != b'\0' and byte != 0 and byte is not None:
        result += chr(byte[0])
        byte = f.read(1)
    
    return f'"{result}"'

def parse(f):
    header = f.read(6)
    assert header == HEADER, "invalid header"

    handlers = {
        TOKEN_STATEMENT: ('statement', read_stmt),
        TOKEN_EXTERNAL: ('external', read_str),
        TOKEN_DATA_I32: ('i32', read_i32),
        TOKEN_DATA_I64: ('i64', read_i64),
        TOKEN_DATA_F32: ('f32', read_f32),
        TOKEN_DATA_F64: ('f64', read_f64),
    }

    token = f.read(1)
    while token is not None and token != 0 and len(token) == 1:
        if token in handlers:
            handler = handlers[token]
            print(f'{handler[0]}: {handler[1](f)}')
        else:
            token_value = int.from_bytes(token, byteorder=sys.byteorder)
            print(f'invalid token identifier: {token_value}')
            exit(1)

        token = f.read(1)

if __name__ == '__main__':
    if not os.path.isfile(input_file):
        print('no such file or directory: {}'.format(input_file))
        exit(1)

    with open(input_file, 'rb') as f:
        parse(f)
