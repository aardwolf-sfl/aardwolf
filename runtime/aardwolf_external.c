#include <stdio.h>

#include "runtime.h"

int main(int argc, char *argv[])
{
    if (argc == 1) {
        aardwolf_write_header();
    } else if (argc == 2) {
        aardwolf_write_external(argv[1]);
        return 0;
    } else {
        fprintf(stderr, "invalid usage of aardwolf_external");
        return 1;
    }
}
