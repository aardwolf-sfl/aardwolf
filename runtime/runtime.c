#include "runtime.h"

#include <stdlib.h>
#include <stdio.h>
#include <stdint.h>
#include <string.h>

#define FILE_FORMAT_VERSION 1

#define ASCII_ZERO 48

// Opened on the first API use. Closed after the process termination.
static FILE * __aardwolf_fd = NULL;

void __write_header(FILE *fd)
{
    fputs("AARD/D", fd);
    fputc(FILE_FORMAT_VERSION + ASCII_ZERO, fd);
}

FILE * __aardwolf_get_fd(void)
{
    if (__aardwolf_fd == NULL) {
        char *dest_dir = getenv("AARDWOLF_DATA_DEST");

        char filename[] = "aard.trace";
        char * filepath;

        // NOTE: sizeof(filename) includes null terminator as well.
        if (dest_dir == NULL) {
            filepath = (char*)malloc(sizeof(filename));
            strcpy(filepath, filename);
        } else {
            size_t destination_length = strlen(dest_dir) + 1;

            filepath = (char*)malloc(destination_length + sizeof(filename));
            memset(filepath, 0, destination_length + sizeof(filename));

            strcpy(filepath, dest_dir);
            filepath[destination_length - 1] = '/';
            strcpy(filepath + destination_length, filename);
        }

#ifndef NO_HEADER
        __aardwolf_fd = fopen(filepath, "w");
#else
        __aardwolf_fd = fopen(filepath, "a");
#endif

        if (__aardwolf_fd == NULL) {
            fprintf(stderr, "Aardwolf error: cannot open %s.\n", filepath);
            free(filepath);
            exit(1);
        }

#ifndef NO_HEADER
        __write_header(__aardwolf_fd);
#endif

        free(filepath);
    }

    return __aardwolf_fd;
}

void __aardwolf_write_data(uint8_t token, void* data, size_t type_size)
{
#ifndef NO_DATA
    FILE *fd = __aardwolf_get_fd();
    fputc(token, fd);
    fwrite(data, type_size, 1, fd);
    fflush(fd);
#endif
}


void aardwolf_write_statement(statement_ref_t id)
{
    __aardwolf_write_data(TOKEN_STATEMENT, &id, sizeof(statement_ref_t));
}

void aardwolf_write_external(const char *external)
{
#ifndef NO_DATA
    FILE *fd = __aardwolf_get_fd();
    fseek(fd, 0, SEEK_END);
    fputc(TOKEN_EXTERNAL, fd);
    fputs(external, fd);
    fputc(0, fd); // null terminator
    fflush(fd);
#endif
}

void aardwolf_write_header()
{
    __write_header(__aardwolf_get_fd());
}

void aardwolf_write_data_unsupported()
{
    __aardwolf_write_data(TOKEN_DATA_UNSUPPORTED, NULL, 0);
}

void aardwolf_write_data_i8(int8_t value)
{
    __aardwolf_write_data(TOKEN_DATA_I8, &value, sizeof(int8_t));
}

void aardwolf_write_data_i16(int16_t value)
{
    __aardwolf_write_data(TOKEN_DATA_I16, &value, sizeof(int16_t));
}

void aardwolf_write_data_i32(int32_t value)
{
    __aardwolf_write_data(TOKEN_DATA_I32, &value, sizeof(int32_t));
}

void aardwolf_write_data_i64(int64_t value)
{
    __aardwolf_write_data(TOKEN_DATA_I64, &value, sizeof(int64_t));
}

void aardwolf_write_data_u8(uint8_t value)
{
    __aardwolf_write_data(TOKEN_DATA_U8, &value, sizeof(uint8_t));
}

void aardwolf_write_data_u16(uint16_t value)
{
    __aardwolf_write_data(TOKEN_DATA_U16, &value, sizeof(uint16_t));
}

void aardwolf_write_data_u32(uint32_t value)
{
    __aardwolf_write_data(TOKEN_DATA_U32, &value, sizeof(uint32_t));
}

void aardwolf_write_data_u64(uint64_t value)
{
    __aardwolf_write_data(TOKEN_DATA_U64, &value, sizeof(uint64_t));
}

void aardwolf_write_data_f32(float value)
{
    __aardwolf_write_data(TOKEN_DATA_F32, &value, sizeof(float));
}

void aardwolf_write_data_f64(double value)
{
    __aardwolf_write_data(TOKEN_DATA_F64, &value, sizeof(double));
}
