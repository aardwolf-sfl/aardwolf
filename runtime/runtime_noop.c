#include "runtime.h"

// An implementation of runtime interface with empty bodies. This may be useful
// for test cases which call `aardwolf_write_external` when just testing the
// code without using Aardwolf analysis.

void aardwolf_write_statement(statement_ref_t id) { }

void aardwolf_write_external(const char *external) { }

void aardwolf_write_data_i32(int32_t value) { }

void aardwolf_write_data_i64(int64_t value) { }

void aardwolf_write_data_f32(float value) { }

void aardwolf_write_data_f64(double value) { }
