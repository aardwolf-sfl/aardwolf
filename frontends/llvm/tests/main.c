#include <stdio.h>
#include "fib.c"

#include "../../../runtime/runtime.h"

int main()
{
    aardwolf_write_external("fib");
    int f = fib(5);
    printf("%d\n", f);
    return 0;
}
