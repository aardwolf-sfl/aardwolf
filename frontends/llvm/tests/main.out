
function: fib

#1 -> #2  ::  defs: %1 / uses:  [@1, 1, 13]
#2 -> #3  ::  defs: %2 / uses:  [@1, 4, 7]
#3 -> #4  ::  defs: %3 / uses:  [@1, 5, 7]
#4 -> #5  ::  defs: %4 / uses:  [@1, 7, 12]
#5 -> #6, #10  ::  defs:  / uses: %4, %1 [@1, 7, 5]
#6 -> #7  ::  defs: %5 / uses: %2, %3 [@1, 8, 11]
#7 -> #8  ::  defs: %2 / uses: %3 [@1, 9, 11]
#8 -> #9  ::  defs: %3 / uses: %5 [@1, 10, 11]
#9 -> #5  ::  defs: %4 / uses: %4 [@1, 7, 25]
#10 ->   ::  defs:  / uses: %3 [@1, 13, 5]

function: main

#11 -> #12  ::  defs:  / uses:  [@2, 8, 5]
#12 -> #13  ::  defs: %6 / uses:  [@2, 9, 13]
#13 -> #14  ::  defs: %7 / uses: %6 [@2, 9, 9]
#14 -> #15  ::  defs: %8 / uses: %7 [@2, 10, 5]
#15 ->   ::  defs:  / uses:  [@2, 11, 5]

@1 = tests/fib.c
@2 = tests/main.c
